### ISSUE TITLE
local {copy,move} custom storage volumes

### ISSUE THREAD

**Issue by brauner** [lxc/lxd#3984]
- local copy custom storage volume in the same pool
- local copy custom storage volumes between pools
- local move custom storage volumes between pools

**Comment by katiewasnothere** [comment]
Hi! Could you further explain this and provide updates if there are any? 

Myself along with a few undergrad university students are looking for a feature project to implement in approximately one month's amount of time. Let me know if this is something we could potentially help contribute to!

**Comment by stgraber** [comment]
So this issue is trying to fix a number of shortcomings in our current storage story.

Right now we allow the creation of storage pools and of custom volumes on those storage pools. We've recently added code to allow renaming existing custom storage volumes but we're missing a few more of the expected operations.

1) Creating a new volume as a copy of an existing one. That'd most likely be used when making a copy of a container which uses custom storage volumes. The expectation is that the copied volume would have the same content and properties as the original.

This would likely look like (assuming a pool called "default" and volume called "foo":

    lxc storage volume copy default/foo bar

2) Similar to 1) but between storage pools. The difficulty there is that storage pools may be using different filesystems so this would need to re-use the same logic we have for container migration, to detect what's the best way to copy the data.

This would look like (assuming a source pool called "ssd", a target pool called "spindle" and a volume called "foo"):

    lxc storage volume copy ssd/foo spindle/foo

3) Same as 2) but since it's a move, the source will then be deleted and any reference to it will be changed to point to the new target. A move would only be allowed if all container using the storage volume are stopped.

This would look like (same example as above but with a move):

    lxc storage volume move ssd/foo spindle/foo

Hope that helps clarify what this issue is tracking!

**Comment by katiewasnothere** [comment]
Can you explain a little about the current storage story? Is the copying mechanism for volumes already in place?

**Comment by stgraber** [comment]
We have a number of functions for every one of the supported storage backends (dir, zfs, btrfs, lvm and ceph) that can likely be re-used for custom volume copy.

Those functions are the ones we use today when copying a container (which results in its container storage volume being copied) or when migrating a container (which involves copying its storage volume to another machine over the network).

For cases where we've not made a clear separation between the low level operation (copying a volume on the filesystem) from the higher level operation (copying a container), there'd need to be a bit of refactoring so that the low level code can be re-used for custom volume copies.

**Comment by jialin-li** [comment]
We started on implementing the copy operation within the same storage pool and have a few quetsions:
1. What would be an appropriate new endpoint for the copy operation? Should we add append to the existing `/1.0/storage-pools/<pool>/volumes/<type>/<name>/` endpoint or should we make a more general copy device endpoint?
2. We are implementing copy for the btrfs for now, and we use the source as `/var/lib/lxd/storage-pools/default/custom/src`, and dst as `/var/lib/lxd/storage-pools/default/custom/dst`. But when we do a btrfs subvolume snapshot, instead of having the data in src copied into dst, we get the entire src folder inside the dst volume.
3. As of now, there is no support for copy operation across storage pool. Should we use migration logic for this scenario?

**Comment by stgraber** [comment]
What we've done in the past for copy (in the case of containers) was to make it part of the create operation with a "source" struct part of the initial creation request struct.

So this would indeed be "POST /1.0/storage-pools/\<pool\>/volumes" which would then take an additional (optional) "source" field that can be used for local or remote copy of a volume.

For copying between two storage pools, some version of the migration code should indeed be used as its source/sink mechanism supports transferring between different storage backends.

I'm not sure about the btrfs problem, maybe @brauner will have some ideas.

**Comment by jialin-li** [comment]
We've figured out the btrfs problem, if dst exist before the snapshot, the snapshot creates a src folder inside dst. If dst doesn't exist before the snapshot, the snapshot just copies and renames src.

**Comment by brauner** [comment]
@jialin-li, what codepath are you looking specifically here? :)

**Comment by jialin-li** [comment]
I added a CopyVolume function to the storage interface, and am calling btrfsSnapshot() in CopyVolume() in storage_btrfs.go. I've fixed the previous problem, but have another question. When we copy the volume, should we set the new volume's config and description the same as the source volume?

**Comment by brauner** [comment]
> When we copy the volume, should we set the new volume's config and description the same as the source volume?

Yes, that's what I would expect (I hope I'm not missing a corner case but in general I expect the properties to be the same.). However, a few things to consider:
- If the source and target storage pool have the same storage type then inheritance is the default for all properties. Specifically:
  - If the source volume has explicit per-volume properties set. (For btrfs this would only include `size`, I think.) then the target volume will inherit this property.
  - If the source volume inherits the properties from pool properties (For btrfs this would only include `size`, I think.) there are two scenarios:
    1. The target pool has the property with the same value set. In this case the target volume will simply inherit the property from the pool as did the source volume.
    2. The target pool has the property with a different value set. In this case the target volume will **not** inherit the property from the pool. Instead it will receive the corresponding volume property. For example, if the source volume inherits its `size` property from the pool via `volume.size` and the target pool has a different `volume.size` property set then the target volume will receive a `size` property set to the value of the source pool's `volume.size` property.
- If the source and target storage pool have different storage types then properties that don't make sense for the receiving pool will simply be dropped. For comparable properties the aforementioned rules apply.

Does that answer your question sufficiently or do you need more detail? :)

**Comment by jialin-li** [comment]
@brauner Thanks! That makes sense, but I don't necessarily see the inheritance of storage pool config to the volumes. Is it an explicit inheritance or implicit due to the size limit of the storage pool?  Also does it make sense to apply storage pool config to the volume? For example, we can set btrfs size to 41943040, but isn't that the capacity of the storage pool? Why would we apply this to the individual volume?

**Comment by brauner** [comment]
> @brauner Thanks! That makes sense, but I don't necessarily see the inheritance of storage pool config to the volumes. Is it an explicit inheritance or implicit due to the size limit of the storage pool? Also

The inheritance is implicit in so far as there's a default value for any configuration key that applies to all volumes in a given pool without being explicitly set to that value. For example, for the lvm storage type the size of newly created volumes is `10GB`. This value is simply assumed. If users want all new volumes to be of a different size they need to set `volume.block.size` to a different value.

 does it make sense to apply storage pool config to the volume? For example, we can set btrfs size to 41943040, but isn't that the capacity of the storage pool? Why would we apply this to the individual volume?

You're mixing up keys that apply to the pool itself (e.g. `size`) with configuration keys that set on the pool level but do only apply to each volume created in the storage pool (e.g. `volume.size`). The difference is that the `size` property for the pool regulates the size of the pool itself whereas the `volume.size` property regulates the size of each volume in the pool. To get a clearer understanding you can look at `lxd/storage_volumes_config.go()`. For example, for the size property of storage volumes we check:

```
		// Does the pool request a default size for new storage volumes?
		if config["size"] == "0" || config["size"] == "" {
			config["size"] = parentPool.Config["volume.size"]
		}
		// Does the user explicitly request a default size for new
		// storage volumes?
		if config["size"] == "0" || config["size"] == "" {
			config["size"] = "10GB"
		}
```

Does that help?

**Comment by jialin-li** [comment]
@brauner Yes! That makes sense :) I am not sure if the inherit case is different from the explicit per-volume case your described. Because a volume that doesn't explicitly specify the config inherits its config from the storage pool, and right now I just copy over the src volume config to dst volume config, which should already reflect the default pool config.

**Comment by brauner** [comment]
> @brauner Yes! That makes sense :) I am not sure if the inherit case is different from the explicit per-volume case you described. Because a volume that doesn't explicitly specify the config inherits its config from the storage pool, and right now I just copy over the src volume config to dst volume config, which should already reflect the parent pool's config.

Yes, that sounds correct.

**Comment by jialin-li** [comment]
@brauner For the ceph driver, the storage pool can specify if it wants a full copy or sparse copy by setting `config["ceph.rbd.clone_copy"]`. But even if the configuration says so, does it make sense to do a sparse copy on a volume? I ask this because it introduces a dependency relationship between source and volume.

**Comment by brauner** [comment]
On Mon, Dec 04, 2017 at 07:37:41PM -0800, Jialin Li wrote:
> @brauner For the ceph driver, the storage pool can specify if it wants a full copy or sparse copy by setting `config["ceph.rbd.clone_copy"]`. But even if the configuration says so, does it make sense to do a sparse copy on a volume? I ask this because it introduces a dependency relationship between source and volume.

I guess you're talking about copying a ceph storage volume within the
same ceph storage pool, right? Because afaik sparse copies between
different ceph storage pools shouldn't be possible.
As long as source and target volume are on the same pool it does make
sense to support sparse copies and would align with what we do for
container copies in the same storage pool. It seems that you already
have a pretty good handle on how the these dependencies are tracked in
the ceph storage driver. It's basically centered around the concept of a
"zombie" volume as you can see from the codebase. :) But feel free to
ask more questions. :)

Christian

**Comment by stgraber** [comment]
Assigning to @brauner now for LXD 3.0.

**Comment by jialin-li** [comment]
@brauner We've done copy within the same storage pool on this branch https://github.com/jialin-li/lxd/tree/storage  Ceph is not done, lvm might have some problems as well, but a lot of the interface work and copy for other storage pools should be working if you don't mind looking over it :)

**Comment by brauner** [comment]
On Thu, Jan 04, 2018 at 10:30:00PM +0000, Jialin Li wrote:
> @brauner We've done copy within the same storage pool on this branch
> https://github.com/jialin-li/lxd/tree/storage  Ceph is not done, lvm

Excellent that you've kept the branch. Please don't delete it. :)

> might have some problems as well, but a lot of the interface work and
> copy for other storage pools should be working if you don't mind
> looking over it :)

Oh sure, that is great news. I was wondering how far you've gotten. I'll
take a look soon (I hope.). I take it that you're not going to finish
this task anymore?

Christian

**Comment by jialin-li** [comment]
@brauner  I would love to continue working on it, but if you guys need it by a certain deadline, depends on the deadline I'm not sure if I can make it. If you don't mind I can keep working on it :)

**Comment by stgraber** [comment]
Looking at other stuff we want for LXD 3.0, we'd be looking at getting this merge by end of January.
Would that work for you?

**Comment by jialin-li** [comment]
@stgraber  We should be able to put up a pr for copy within the same pool soon after we write some tests. For copy and move between the storage pool I will work on it and let you know if it doesn't seem to be on track for the deadline. Is that okay?

**Comment by stgraber** [comment]
@jialin-li sounds good, thanks!

**Comment by jialin-li** [comment]
@stgraber It doesn't seem like I will have enough time to work on copy and move between storage pool, will have a pr on what we have so far soon! Sorry, on vacation right now, don't have much time to work :(

**Comment by stgraber** [comment]
@jialin-li Sounds good, we'll either merge what you have and do the rest ourselves or we'll push the missing bits directly to your branch and then review+merge. Thanks for all the work!
